internal/v5/archive.go

```diff
@@ -333,6 +334,17 @@ func (h *ReqHandler) serveArchiveFile(id *router.ResolvedURL, w http.ResponseWri
 // with the given id.
 func (h *ReqHandler) ServeBlobFile(w http.ResponseWriter, req *http.Request, id *router.ResolvedURL, blob *charmstore.Blob) error {
 	r, size, err := h.Store.OpenBlobFile(blob, req.URL.Path)
+	if redirect, ok := err.(*charmstore.ErrRedirect); ok {
+		p, err := router.RelativeURLPath(req.URL.Path, path.Join(path.Dir(req.URL.Path), redirect.Path))
+		if err != nil {
+			return errgo.Notef(err, "cannot make relative URL from %q and %q", req.URL.Path, path.Join(path.Dir(req.URL.Path), redirect.Path))
+		}
+		// We can't use the http.redirect function because it tries to build an absolute url but the path in the
+		// request URL has been changed.
+		w.Header().Set("Location", p)
+		w.WriteHeader(http.StatusMovedPermanently)
```

Comment:
I think StatusFound would be better. We cannot say it is moved perminently only until a new zip file is uploaded.